src/gluonts/core/transformation_check.py

```diff
@@ -0,0 +1,114 @@
+    return this.__code__.co_code == that.__code__.co_code
+
+
+def equals_init_args(this: Transformation, that: Transformation) -> bool:
+
+    if type(this) != type(that):
+        return False
+
+    this_args = inspect.signature(this.__init__).parameters.keys()
+    that_args = inspect.signature(that.__init__).parameters.keys()
+
+    if set(this_args) != set(that_args):
+        return False
+
+    for arg in this_args:
+        if callable(getattr(this, arg)):
+            equal_func = equals_callable
+        elif isinstance(getattr(this, arg), InstanceSampler):
+            equal_func = equals_instance_sampler
+        else:
+            equal_func = equals
+
+        if not equal_func(getattr(this, arg), getattr(that, arg)):
+            return False
+
+    return True
+
+
+@equals.register(Transformation)
+def equals_transformation(this: Transformation, that: Transformation):
+    return equals_init_args(this, that)
+
+
+@equals.register(MapTransformation)
+def equals_map_transformation(this: MapTransformation, that: MapTransformation):
+
+    if not equals_transformation(this, that):
+        return False
+
+    if not equals_callable(this.map_transform, that.map_transform):
+        return False
+
+    return True
+
+
+@equals.register(SimpleTransformation)
+def equals_simple_transformation(this: SimpleTransformation, that: SimpleTransformation):
+
+    if not equals_map_transformation(this, that):
+        return False
+
+    if not equals_callable(this.transform, that.transform):
+        return False
+
+    return True
+
+
+@equals.register(FlatMapTransformation)
+def equals_flatmap_transformation(this: FlatMapTransformation, that: FlatMapTransformation):
+
+    if not equals_transformation(this, that):
+        return False
+
+    if not equals_callable(this.flatmap_transform, that.flatmap_transform):
+        return False
+
+    return True
+
+
+@equals.register(Batch)
+def equals_batch(this: Batch, that: Batch):
+
+    if not equals_transformation(this, that):
+        return False
+
+    if this.batch_size != that.batch_size:
+        return False
+
+    return True
+
+
+@equals.register(Chain)
+def equal_chain(this: Chain, that: Chain):
+
+    if len(this.transformations) != len(that.transformations):
+        return False
+
+    for this_tran, that_tran in zip(this.transformations, that.transformations):
+        # we cannot call equals here, otherwise, reach maximum recursion
+        if isinstance(this_tran, Batch):
+            equals_batch(this_tran, that_tran)
+        elif isinstance(this_tran, FlatMapTransformation):
+            equals_flatmap_transformation(this_tran, that_tran)
+        elif isinstance(this_tran, SimpleTransformation):
+            equals_simple_transformation(this_tran, that_tran)
+        elif isinstance(this_tran, MapTransformation):
+            equals_map_transformation(this_tran, that_tran)
+        else:
+            equals_transformation(this_tran, that_tran)
```

Comment:
- Isn't a stopping criteria needed to stop the for-loop and return false if any of the equals returns false?
- Do you have a minimal example of 2 chains where the maximum recursion depth is reached?

Reply:
Thanks for the comments!

* Re the early stop the for loop, absolutely, will add that in the next revision, thanks for catching it!
* Re the  recursion, because we decorate the function by "equals", that's why if we call "equals" inside this function, it will certainly reach maximum recursion. Will provide tests in the next revision.

Reply:
Chances are that I have missed something here. That's why I asked for an example.
`equals` is not a function but a generic-function with singledispatch (kind of a set of functions). That means, `equals` forwards (dispatches) its arguments to the specific function based on the type of the first argument. The complete body of your for-loop is equivalent to calling just `equals(this_tran, that_tran)`. You can try this out. 
It would be something like that:
```python
for this_tran, that_tran in zip(this.transformations, that.transformations):
    if not equals(this_tran, that_tran):
        return False
```
I hope I haven't missed anything.

Reply:
Ah, you are right! I confused myself with a wrong example (where the setting is a bit different with inheritance). This is much better indeed!